File: backend/app/services/risk.py
from app.models.vault import Vault, RiskFactors, RiskAnalysis



def calculate_protocol_dependency_risk(dependencies: list[str]) -> float:
    if len(dependencies) == 0:
        return 2.0
    elif len(dependencies) == 1:
        return 7.0

def calculate_orac
        "chainlink": 2.0,
        "internal"
    }


def calculate_oracle_risk(oracle_type: str) -> float:
    oracle_scores = {
        "chainlink": 2.0,
        "uniswap": 5.0,
        "internal": 8.0,
        "none": 1.0,
    }
    return oracle_scores.get(oracle_type.lower(), 5.0)


def calculate_upgradeability_risk(upgradeability: str) -> float:
    upgrade_scores = {
        "immutable": 1.0,
        "timelock": 3.0,
        "multisig": 6.0,
        "eoa": 10.0,
    }
    return upgrade_scores.get(upgradeability.lower(), 5.0)


def calculate_liquidity_risk(liquidity_depth: float, tvl: float) -> float:
    if tvl == 0:
        return 9.0
    
    ratio = liquidity_depth / tvl
    if ratio >= 3:
        return 2.0
    elif ratio >= 2:
        return 3.0
    elif ratio >= 1.5:
        return 5.0
    elif ratio >= 1:
        return 7.0
    return 9.0


def get_risk_level(score: float) -> Literal["low", "medium", "high"]:
    if score <= 3:
        return "low"
    elif score <= 6:
        return "medium"
    return "high"


def calculate_risk_score(vault: Vault) -> RiskAnalysis:
    protocol_dependency = calculate_protocol_dependency_risk(vault.dependencies)
    oracle_risk = calculate_oracle_risk(vault.oracle_type)
    upgradeability_risk = calculate_upgradeability_risk(vault.upgradeability)
    liquidity_risk = calculate_liquidity_risk(vault.liquidity_depth, vault.tvl)

    weights = {
        "protocol_dependency": 0.30,
        "oracle_risk": 0.25,
        "upgradeability_risk": 0.25,
        "liquidity_risk": 0.20,
    }

    overall_score = (
        protocol_dependency * weights["protocol_dependency"]
        + oracle_risk * weights["oracle_risk"]
        + upgradeability_risk * weights["upgradeability_risk"]
        + liquidity_risk * weights["liquidity_risk"]
    )

    overall_score = round(overall_score, 1)

    return RiskAnalysis(
        vault_id=vault.id,
        overall_score=overall_score,
        level=get_risk_level(overall_score),
        factors=RiskFactors(
            protocol_dependency=protocol_dependency,
            oracle_risk=oracle_risk,
            upgradeability_risk=upgradeability_risk,
            liquidity_risk=liquidity_risk,
        ),
    )

def calculate_upgradeability_risk(upgradeability: str) -> float:
    upgrade_scores = {
        "immutable": 1.0,
        "timelock": 3.0,
        "multisig": 6.0,
        "eoa": 10.0,
    }
    return upgrade_scores.get(upgradeability.lower(), 5.0)


def calculate_liquidity_risk(liquidity_depth: float, tvl: float) -> float:
    if tvl == 0:
        return 9.0
    
    ratio = liquidity_depth / tvl
    if ratio >= 3:
        return 2.0
    elif ratio >= 2:
        return 3.0
    elif ratio >= 1.5:
        return 5.0
    elif ratio >= 1:
        return 7.0
    return 9.0


def get_risk_level(score: float) -> Literal["low", "medium", "high"]:
    if score <= 3:
        return "low"
    elif score <= 6:
        return "medium"
    return "high"


def calculate_risk_score(vault: Vault) -> RiskAnalysis:
    protocol_dependency = calculate_protocol_dependency_risk(vault.dependencies)
    oracle_risk = calculate_oracle_risk(vault.oracle_type)
    upgradeability_risk = calculate_upgradeability_risk(vault.upgradeability)
    liquidity_risk = calculate_liquidity_risk(vault.liquidity_depth, vault.tvl)

    weights = {
        "protocol_dependency": 0.30,
        "oracle_risk": 0.25,
        "upgradeability_risk": 0.25,
        "liquidity_risk": 0.20,
    }

    overall_score = (
        protocol_dependency * weights["protocol_dependency"]
        + oracle_risk * weights["oracle_risk"]
        + upgradeability_risk * weights["upgradeability_risk"]
        + liquidity_risk * weights["liquidity_risk"]
    )

    overall_score = round(overall_score, 1)

    return RiskAnalysis(
        vault_id=vault.id,
        overall_score=overall_score,
        level=get_risk_level(overall_score),
        factors=RiskFactors(
            protocol_dependency=protocol_dependency,
            oracle_risk=oracle_risk,
            upgradeability_risk=upgradeability_risk,
            liquidity_risk=liquidity_risk,
        ),
    )

File: src/lib/risk.ts
import { RiskFactors, Vault } from './types';

export function calculateRiskScore(vault: Vault): { score: number; factors: RiskFactors } {
  const protocolDependency = calculateProtocolDependencyRisk(vault.dependencies);
  const oracleRisk = calculateOracleRisk(vault.oracleType);
  const upgradeabilityRisk = calculateUpgradeabilityRisk(vault.upgradeability);
  const liquidityRisk = calculateLiquidityRisk(vault.liquidityDepth, vault.tvl);

  const weights = {
    protocolDependency: 0.30,
    oracleRisk: 0.25,
    upgradeabilityRisk: 0.25,
    liquidityRisk: 0.20,
  };

  const score =
    protocolDependency * weights.protocolDependency +
    oracleRisk * weights.oracleRisk +
    upgradeabilityRisk * weights.upgradeabilityRisk +
    liquidityRisk * weights.liquidityRisk;

  return {
    score: Math.round(score * 10) / 10,
    factors: {
      protocolDependency,
      oracleRisk,
      upgradeabilityRisk,
      liquidityRisk,
    },
  };
}

function calculateProtocolDependencyRisk(dependencies: string[]): number {
  if (dependencies.length === 0) return 2;
  if (dependencies.length === 1) return 3;
  if (dependencies.length === 2) return 5;
  if (dependencies.length === 3) return 7;
  return 9;
}

function calculateOracleRisk(oracleType: Vault['oracleType']): number {
  switch (oracleType) {
    case 'chainlink':
      return 2;
    case 'uniswap':
      return 5;
    case 'internal':
      return 8;
    case 'none':
      return 1;
  }
}

function calculateUpgradeabilityRisk(upgradeability: Vault['upgradeability']): number {
  switch (upgradeability) {
    case 'immutable':
      return 1;
    case 'timelock':
      return 3;
    case 'multisig':
      return 6;
    case 'eoa':
      return 10;
  }
}

function calculateLiquidityRisk(liquidityDepth: number, tvl: number): number {
  const ratio = liquidityDepth / tvl;
  if (ratio >= 3) return 2;
  if (ratio >= 2) return 3;
  if (ratio >= 1.5) return 5;
  if (ratio >= 1) return 7;
  return 9;
}

export function getRiskLevel(score: number): 'low' | 'medium' | 'high' {
  if (score <= 3) return 'low';
  if (score <= 6) return 'medium';
  return 'high';
}

export function getRiskColor(score: number): string {
  const level = getRiskLevel(score);
  switch (level) {
    case 'low':
      return 'oklch(0.70 0.20 145)';
    case 'medium':
      return 'oklch(0.75 0.18 85)';
    case 'high':
      return 'oklch(0.65 0.24 25)';
  }
}

export function formatCurrency(value: number): string {
  if (value >= 1_000_000_000) {
    return `$${(value / 1_000_000_000).toFixed(2)}B`;
  }
  if (value >= 1_000_000) {
    return `$${(value / 1_000_000).toFixed(2)}M`;
  }
  if (value >= 1_000) {
    return `$${(value / 1_000).toFixed(2)}K`;
  }
  return `$${value.toFixed(2)}`;
}

export function formatPercent(value: number, decimals: number = 2): string {
  return `${value.toFixed(decimals)}%`;
}

export function formatAddress(address: string): string {
  return `${address.slice(0, 6)}...${address.slice(-4)}`;
}

export function generateYieldHistory(baseAPY: number, days: number = 90): Array<{ timestamp: number; apy: number; tvl: number }> {
  const data = [];
  const now = Date.now();
  const dayMs = 24 * 60 * 60 * 1000;
  
  for (let i = days; i >= 0; i--) {
    const variance = (Math.random() - 0.5) * baseAPY * 0.2;
    const apy = Math.max(0, baseAPY + variance);
    const tvlVariance = (Math.random() - 0.5) * 0.3;
    const tvl = Math.max(0, 1 + tvlVariance);
    
    data.push({
      timestamp: now - i * dayMs,
      apy,
      tvl,
    });
  }
  
  return data;
}
